include/util/dynamic_graph.hpp

```diff
@@ -309,6 +332,36 @@ template <typename EdgeDataT> class DynamicGraph
         return current_iterator;
     }
 
+    void Renumber(const std::vector<NodeID> &old_to_new_node)
+    {
+        // permutate everything but the sentinel
+        util::inplacePermutation(node_array.begin(), std::prev(node_array.end()), old_to_new_node);
+
+        // Build up edge permutation
+        auto new_edge_index = 0;
+        std::vector<EdgeID> old_to_new_edge(edge_list.size(), SPECIAL_EDGEID);
+        for (auto node : util::irange<NodeID>(0, number_of_nodes))
+        {
+            auto new_first_edge = new_edge_index;
+            // move all filled edges
+            for (auto edge : GetAdjacentEdgeRange(node))
+            {
+                edge_list[edge].target = old_to_new_node[edge_list[edge].target];
+                old_to_new_edge[edge] = new_edge_index++;
+            }
+            // and all adjacent empty edges
+            for (auto edge = EndEdges(node); edge < number_of_edges && isDummy(edge); edge++)
```

Comment:
I mainly wanted the renumber operation to not interfere with the structure of the dynamic graph. Removing the dummy edges would compactify the structure but would have non-obvious side-effects like for example `InsertEdge` would now take `O(n)` because it needs to resize+copy the edge array and relocate the source node edges.

We might want an additional function `shrink_to_fit` or something like that, but I don't think we have good use for it right now.